### DISCUSSION TITLE
How to print metric value every epoch ?

### DISCUSSION THREAD

**Discussion by richarddwang** [Lightning-AI/pytorch-lightning#6407] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
**What I want it to look like:**
![image](https://user-images.githubusercontent.com/17963619/110324632-3253ef80-8051-11eb-8ec0-abe20b9100e0.png)

**What I got now:**
![image](https://user-images.githubusercontent.com/17963619/110324679-4566bf80-8051-11eb-9518-cf4b317edc5f.png)

**I think I do have logged metrics, how to make it printed beautifully ?**
```
class SortNumberModel(pl.LightningModule):
    def __init__(self, hf_config, lr):
        super().__init__()
        self.save_hyperparameters()
        self.model = AutoModelForTokenClassification.from_config(hf_config)
        self.val_acc = pl.metrics.Accuracy()

    def forward(self, batch):
        result = self.model(input_ids=batch["x"], labels=batch["y"], return_dict=True)
        return result.logits.argmax(dim=-1), result.loss

    def training_step(self, batch, batch_idx):
        return self.forward(batch)[-1]

    def validation_step(self, batch, batch_idx):
        preds, loss = self.forward(batch)
        self.val_acc(preds.view(-1), batch["y"].view(-1))
        self.log("valid_acc", self.val_acc, on_epoch=True)

    def configure_optimizers(self):
        return torch.optim.AdamW(self.parameters(), lr=self.hparams.lr, eps=1e-5)
```

**Comment by SkafteNicki** [comment, marked as answer]
In pytorch-lightning bolts we have a few extra callbacks, with one of them being the `PrintTableMetricsCallback` callback that should output exactly what you want. Simply install bolts as 
``` python
pip install pytorch-lightning-bolts
```
and initialize your trainer as:
```python
trainer = Trainer(callbacks=[PrintTableMetricsCallback()])
```

**Reply by richarddwang** [reply]
Thanks for your comment.
But I got this in vscode notebook for
```
trainer = pl.Trainer(
    max_epochs=1, gpus="0", callbacks=[PrintTableMetricsCallback()]
)
trainer.fit(model, datamodule=dm)
```
![image](https://user-images.githubusercontent.com/17963619/110399386-bee1ca80-80b0-11eb-94e0-0cf9c41b0536.png)

It seems that it even prints for valid step, any good suggestion ?

**Reply by SkafteNicki** [reply]
I can see that the callback is not doing exactly what you want.
I tried to type up another callback that should do something closer to what you want:
```python
from pytorch_lightning.callbacks import Callback
from copy import deepcopy
import tableprint as tp

class Printer(Callback):
    def __init__(self) -> None:
        self.metrics = {}
        self.table_context = None

    def on_train_epoch_end(self, trainer, pl_module, outputs) -> None:
        metrics_dict = deepcopy(trainer.callback_metrics)
        if self.table_context is None:
            self.table_context = tp.TableContext(headers=metrics_dict.keys())
            self.table_context.__enter__()
        self.table_context([v.item() for v in metrics_dict.values()])

    def on_train_end(self, trainer, pl_module):
        self.table_context.__exit__()
```
this callback requires that you install https://github.com/nirum/tableprint. With this, it should output something like this to the terminal:
![snip](https://user-images.githubusercontent.com/24896311/110641145-eec4c700-81b1-11eb-956a-65d2fbc5a574.PNG)

**Reply by richarddwang** [reply]
Thanks for following up !

I think to print validation accuracy/loss, it may be more reasonable if we add info to the table at `on_validation_epoch_end`.
Otherwise it seems to print metrics calculated at previous validation loop.
So I made a little modification.
```python
class Printer(Callback):
    def __init__(self) -> None:
        self.metrics = {}
        self.table_context = None

    def on_validation_epoch_end(self, trainer, pl_module) -> None:
        if trainer.running_sanity_check:
            return
        metrics_dict = deepcopy(trainer.callback_metrics)
        if self.table_context is None:
            self.table_context = tp.TableContext(headers=metrics_dict.keys())
            self.table_context.__enter__()
        self.table_context([v.item() for v in metrics_dict.values()])

    def on_fit_end(self, trainer, pl_module):
        self.table_context.__exit__()
```
(Cuz sanity check also calls validation loop and thus activate `on_validaton_epoch_end`, I also add an if clause to skip it.)

And it now pretty fit my need.
![image](https://user-images.githubusercontent.com/17963619/110762274-94983480-828b-11eb-8a12-7cb4fbb31862.png)

I have tried pt for few days, it is really neat. It will be nice if someone can contribute an default(or off-the-shelf) and non-novel table printing to it.

**Comment by rudaoshi** [comment]
@SkafteNicki,  Thank you for you answer. 

However, I found the solution prints different metric values accross different ddp process. 

I'm struggling to correctly print the metrics on the end of each epoch, but I meet a lot of problem discribed in this issue:

https://github.com/PyTorchLightning/pytorch-lightning/issues/6122#issuecomment-797863438 

Will you please look through my problems and give some suggestions? 

Thank you for your help!
